lib/oscap_tests.pm

```diff
@@ -0,0 +1,348 @@
+# Priority Medium:
+our $profile_ID_sle_anssi_bp28_enhanced = 'xccdf_org.ssgproject.content_profile_anssi_bp28_enhanced';
+our $profile_ID_sle_cis_server_l1 = 'xccdf_org.ssgproject.content_profile_cis_server_l1';
+our $profile_ID_sle_cis_workstation_l2 = 'xccdf_org.ssgproject.content_profile_cis_workstation_l2';
+# Priority Low:
+our $profile_ID_sle_anssi_bp28_intermediary = 'xccdf_org.ssgproject.content_profile_anssi_bp28_intermediary';
+our $profile_ID_sle_anssi_bp28_minimal = 'xccdf_org.ssgproject.content_profile_anssi_bp28_minimal';
+our $profile_ID_sle_cis_workstation_l1 = 'xccdf_org.ssgproject.content_profile_cis_workstation_l1';
+our $profile_ID_tw = 'xccdf_org.ssgproject.content_profile_standard';
+
+# The OS status of remediation: '0', not remediated; '1', remediated
+our $remediated = 0;
+
+# Upload HTML report by default
+set_var('UPLOAD_REPORT_HTML', 1);
+
+# Set value for 'scap-security-guide' ds file
+sub set_ds_file {
+
+    # Set the ds file for separate product, e.g.,
+    # for SLE15 the ds file is "ssg-sle15-ds.xml";
+    # for SLE12 the ds file is "ssg-sle12-ds.xml";
+    # for Tumbleweed the ds file is "ssg-opensuse-ds.xml"
+    my $version = get_required_var('VERSION') =~ s/([0-9]+).*/$1/r;
+    $f_ssg_sle_ds =
+      '/usr/share/xml/scap/ssg/content/ssg-sle' . "$version" . '-ds.xml';
+}
+
+sub upload_logs_reports {
+
+    # Upload logs & ouputs for reference
+    my $files;
+    if (is_sle) {
+        $files = script_output('ls | grep "^ssg-sle.*.xml"');
+    }
+    else {
+        $files = script_output('ls | grep "^ssg-opensuse.*.xml"');
+    }
+    foreach my $file (split("\n", $files)) {
+        upload_logs("$file");
+    }
+    upload_logs("$f_stdout") if script_run "! [[ -e $f_stdout ]]";
+    upload_logs("$f_stderr") if script_run "! [[ -e $f_stderr ]]";
+    if (get_var('UPLOAD_REPORT_HTML')) {
+        upload_logs("$f_report", timeout => 600)
+          if script_run "! [[ -e $f_report ]]";
+    }
+}
+
+sub pattern_count_in_file {
+
+    #Find count and rules names of matched pattern
+    my $self = $_[0];
+    my $data = $_[1];
+    my $pattern = $_[2];
+    my @rules;
+    my $count = 0;
+
+    my @lines = split /\n|\r/, $data;
+    for my $i (0 .. $#lines) {
+        if ($lines[$i] =~ /$pattern/) {
+            $count++;
+            push(@rules, $lines[$i - 4]);
+        }
+    }
+
+    #Returning by reference array of matched rules
+    $_[3] = \@rules;
+    return $count;
+}
+
+sub rules_count_in_file {
+
+    #Find count of rules names matched name and status patterns
+    my $self = $_[0];
+    my $data = $_[1];
+    my $pattern = $_[2];
+    my $l_rules = $_[3];
+    my @rules;
+    my $count = 0;
+
+    my @lines = split /\n|\r/, $data;
+    my @a_rules = @$l_rules;
+
+    for my $i (0 .. $#lines) {
+        for my $j (0 .. $#a_rules) {
+            if ($lines[$i] =~ /$a_rules[$j]/ and $lines[$i + 4] =~ /$pattern/) {
```

Comment:
- quadratic complexity ?
- second part of the condition is constant in the inner loop 
- consider using Perl grep function

Reply:
It is simplest that I can do when looking for the rule name and checking it passed or failed:
Rule
	xccdf_org.ssgproject.content_rule_is_fips_mode_enabled  >> $lines[$i]
Ident
	CCE-85763-1
Result
	fail                                                                                       >>$lines[$i + 4]

Reply:
Did not see simpler implementation. I tried had to implement grep with pattern - looks OpenQA has issues with it...

Reply:
just to clarify: I'm talking about [Perl's grep function](https://perldoc.perl.org/functions/grep) , not the unix grep(1) command :)

Reply:
grep function you mentioned does not fit the purpose - "actually modifies the element in the original list. This is usually something to be avoided when writing clear code."